pkg/operators/oci/oci.go

```diff
@@ -0,0 +1,275 @@
+			TypeHint:    params.TypeString,
+		},
+	}
+}
+
+func (o *ociHandler) Instantiate(gadgetCtx operators.GadgetContext, params *params.Params) (*OciHandlerInstance, error) {
+	if len(gadgetCtx.ImageName()) == 0 {
+		return nil, fmt.Errorf("imageName empty")
+	}
+
+	secretBytes := []byte{} // TODO
+
+	authOpts := &oci.AuthOptions{
+		AuthFile:    params.Get(authfileParam).AsString(),
+		SecretBytes: secretBytes,
+		Insecure:    params.Get(insecureParam).AsBool(),
+	}
+
+	// Make sure the image is available, either through pulling or by just accessing a local copy
+	// TODO: add security constraints (e.g. don't allow pulling - add GlobalParams for that)
+	err := oci.EnsureImage(gadgetCtx.Context(), gadgetCtx.ImageName(), authOpts, params.Get(pullParam).AsString())
+	if err != nil {
+		return nil, fmt.Errorf("insuring image: %w", err)
+	}
+
+	manifest, err := oci.GetManifestForHost(gadgetCtx.Context(), gadgetCtx.ImageName())
+	if err != nil {
+		return nil, fmt.Errorf("getting manifest: %w", err)
+	}
+
+	logger := gadgetCtx.Logger()
+	logger.Debugf("ArtifactType: %s", manifest.ArtifactType)
+
+	// metadata := &types.GadgetMetadata{}
+	r, err := oci.GetContentFromDescriptor(gadgetCtx.Context(), manifest.Config)
+	if err != nil {
+		return nil, fmt.Errorf("getting metadata: %w", err)
+	}
+	metadata, _ := io.ReadAll(r)
+	r.Close()
+
+	// Store metadata for serialization
+	gadgetCtx.SetMetadata(metadata)
+
+	viper := viper.New()
+	viper.SetConfigType("yaml")
+	err = viper.ReadConfig(bytes.NewReader(metadata))
+
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling metadata: %w", err)
+	}
+
+	gadgetCtx.SetVar("config", viper)
+
+	instance := &OciHandlerInstance{
+		gadgetCtx: gadgetCtx,
+	}
+
+	for _, layer := range manifest.Layers {
+		logger.Debugf("layer > %+v", layer)
+		ops := operators.GetImageOperatorsForMediaType(layer.MediaType)
+		for _, op := range ops {
+			logger.Debugf("found layer op %q", op.Name())
+			opInst, err := op.InstantiateImageOperator(gadgetCtx, layer)
+			if err != nil {
+				logger.Errorf("instantiating operator %q: %v", op.Name(), err)
+			}
+			if opInst == nil {
+				logger.Debugf("> skipped %s", op.Name())
+				continue
+			}
+			instance.layerOperatorInstances = append(instance.layerOperatorInstances, opInst)
+		}
+	}
+
+	for _, op := range operators.GetDataOperators() {
+		logger.Debugf("found data op %q", op.Name())
+
+		// Lazily initialize operator
+		err := op.Init(op.GlobalParamDescs().ToParams())
+		if err != nil {
+			return nil, fmt.Errorf("initializing operator %q: %w", op.Name(), err)
+		}
+
+		opInst, err := op.InstantiateDataOperator(gadgetCtx)
+		if err != nil {
+			logger.Errorf("instantiating operator %q: %v", op.Name(), err)
+		}
+		if opInst == nil {
+			logger.Debugf("> skipped %s", op.Name())
+			continue
+		}
+		instance.dataOperatorInstances = append(instance.dataOperatorInstances, opInst)
+	}
+
+	// Sort dataOperators based on their priority
```

Comment:
The order of operators is important because that is the order in which DataSources are registered. Subscription order just determines the order in which events are forwarded to the operator.